src/middleware/error-handler.ts

```diff
@@ -0,0 +1,65 @@
+import { Request, Response, NextFunction } from "express";
+import CustomAPIError from "../errors/custom-api"; 
+import { StatusCodes } from "http-status-codes";
+import mongoose from "mongoose"; 
+
+interface CustomError {
+  statusCode: number;
+  msg: string;
+}
+
+interface MongoErrorWithKeyValue extends mongoose.mongo.MongoError {
+  keyValue: { [key: string]: any };
+}
+
+export const errorHandlerMiddleware = (
+  err: Error,
+  req: Request,
+  res: Response,
+  next: NextFunction
+): void => {
+  let customError: CustomError = {
+    statusCode: (err as any).statusCode || StatusCodes.INTERNAL_SERVER_ERROR,
+    msg:
+      (err as any).message || "Something went wrong. Please try again later.",
+  };
+
+
+  if (err instanceof mongoose.Error.ValidationError) {
+    customError.statusCode = StatusCodes.BAD_REQUEST;
+    customError.msg = Object.values(err.errors)
+      .map((val: any) => val.message)
+      .join(", ");
+    res.status(customError.statusCode).json({ msg: customError.msg });
+    return;
+  }
+
+  if (err instanceof mongoose.Error.CastError) {
+    customError.statusCode = StatusCodes.BAD_REQUEST;
+    customError.msg = `Invalid value for ${err.path}: ${err.value}`;
+    res.status(customError.statusCode).json({ msg: customError.msg });
+    return;
+  }
+
+  if (err instanceof mongoose.mongo.MongoError && err.code === 11000) {
```

Comment:
What is 11000? That is a magic number to me the reader. Can we make a `const` variable with a descriptive name so that it's clear why we chose that specific number?

Reply:
fixed